### PR TITLE
[#4202] Detect balance and erc20 transfers by a single contract call

### PR DESCRIPTION
**Pull request by rasom**
closes (maybe) #4202

A single call to check them all (almost). 

previous algorithm per iteration per network:
- fetching current block number 
- fetching balance/nonce per account
- 3 getLogs 
- 2 mins interval

new algorithm per network:
- every 2 mins: single contract call to get block number and detect balance changes (including ERC20/721 from the predefined list)
- every 10 mins: 3 getLogs requests to get all erc20/1155/721 which were not detected by previous call
- every 1h: get balances/nonces to check for possible zero sum transfers, per account 

status: ready

### CODE REVIEW

File: services/wallet/transfer/commands_sequential.go [hunk 1]
```diff
@@ -68,7 +72,130 @@ func (c *findNewBlocksCommand) Command() async.Command {
...
+		log.Error("findNewBlocksCommand error creating balance checker", "error", err, "chain", c.chainClient.NetworkID())
+		return nil, nil, err
+	}
+
+	tokens, err := c.tokenManager.GetTokens(c.chainClient.NetworkID())
+	if err != nil {
+		return nil, nil, err
+	}
+	tokenAddresses := []common.Address{}
+	nilAddress := common.HexToAddress("0x0000000000000000000000000000000000000000")
```
Comment (IvanBelyakoff) [inline services/wallet/transfer/commands_sequential.go]:
```suggestion
	nilAddress := common.Address{}
```

File: services/wallet/transfer/commands_sequential.go [hunk 2]
```diff
@@ -68,7 +72,130 @@ func (c *findNewBlocksCommand) Command() async.Command {
...
+		log.Error("findNewBlocksCommand error on transfer detection", "error", err, "chain", c.chainClient.NetworkID())
+		return err
+	}
+
+	if len(accountsWithDetectedChanges) == 0 {
+		// check nonces once per 30 iterations
+		if c.iteration%30 == 0 {
+			c.findAndSaveEthBlocks(parent, c.fromBlockNumber, headNum, accounts)
+		}
+		// check incoming erc20 once per 3 iterations
```
Comment (IvanBelyakoff) [inline services/wallet/transfer/commands_sequential.go]: 5?

File: services/wallet/transfer/commands_sequential.go [hunk 3]
```diff
@@ -68,7 +72,130 @@ func (c *findNewBlocksCommand) Command() async.Command {
...
+		if c.iteration%5 == 0 {
+			c.findAndSaveTokenBlocks(parent, c.fromBlockNumber, headNum)
+		}
+
+		c.fromBlockNumber = headNum
+		c.iteration++
+		return nil
+	}
+
+	c.findAndSaveEthBlocks(parent, c.fromBlockNumber, headNum, accountsWithDetectedChanges)
```
Comment (IvanBelyakoff) [inline services/wallet/transfer/commands_sequential.go]: Up to you, but this feels simpler to me:

```
if len(accountsWithDetectedChanges) != 0  || if c.iteration%30 == 0 {
    c.findAndSaveEthBlocks(parent, c.fromBlockNumber, headNum, accounts)
}
if len(accountsWithDetectedChanges) != 0  || if c.iteration%5 == 0 {
    c.findAndSaveTokenBlocks(parent, c.fromBlockNumber, headNum)
}
c.fromBlockNumber = headNum
c.iteration++
```

File: services/wallet/transfer/commands_sequential.go [hunk 4]
```diff
@@ -68,7 +72,130 @@ func (c *findNewBlocksCommand) Command() async.Command {
...
+	c.findAndSaveTokenBlocks(parent, c.fromBlockNumber, headNum)
+	c.fromBlockNumber = headNum
+	c.iteration++
+
+	return nil
+}
+
+func (c *findNewBlocksCommand) RunOld(parent context.Context) (err error) {
```
Comment (IvanBelyakoff) [inline services/wallet/transfer/commands_sequential.go]: Is it ever called?

**Review by IvanBelyakoff — COMMENTED** [review]
Great job, Roman!
I have some questions to clarify.
So for now we check ETH transfers with 0 sum once an hour (nonces) and possible  missed token transfers (ERC20/721/1155) once in 10 minutes - those that are not whitelisted (only ERC20 are whitelisted) correct?
Please share how will we take care of community token transfers?

File: services/wallet/transfer/commands_sequential.go (hunk 3 above)
Comment (rasom) [reply services/wallet/transfer/commands_sequential.go]: thanks

**Comment by rasom** [comment]
@IvanBelyakoff comments are addressed

**Comment by rasom** [comment]
@IvanBelyakoff 

> So for now we check ETH transfers with 0 sum once an hour (nonces) and possible missed token transfers (ERC20/721/1155) once in 10 minutes - those that are not whitelisted (only ERC20 are whitelisted) correct?

yes, correct, and also I will separately remove nonce check at all for all generated addresses for which mnemonic has not been revealed before.

> Please share how will we take care of community token transfers?

I was told that we receive some metadata with token addresses per community, considering ERC20 and ERC721 have same signature for `balanceOf(owner)` method we would just need to add those addresses to the call which generates hash of balances per account. Then those will be checked once per 2 mins.

**Review by IvanBelyakoff — APPROVED** [review]
LGTM

**Comment by rasom** [comment]
1h, 1 network, 6 accounts (restored from mnemonic) 
```
eth_CallContract_0x08A8fDBddc160A7d5b957256b903dCAb1aE512C5 90 <- ethscan contract main net
eth_CallContract_0xBB85398092B83A016935a17Fc857507b7851a071 90 <- ethscan contract arbitrum
eth_CallContract_0x9e5076DF494FC949aBc4461F4E57592B81517D81 90 <- ethscan contract optimism 
eth_CallContract_0xA8d270048a086F5807A8dc0a9ae0e96280C41e3A 60 <- directory
eth_CallContract_0x040EA8bFE441597849A9456182fa46D38B75BC05 30 <- balance checker (tx detection)
eth_CodeAt                                                  18 <- to be identified
eth_BlockNumber                                             18 <- fetchLatestBlockNumberCommand (to be removed)
eth_FilterLogs                                              18 <- checking incoming erc20/721/1155 (tx detection)
eth_BalanceAt                                               12 <- checking nonce (tx detection) (can be removed)
eth_NonceAt                                                 6  <- checking nonce (tx detection)
eth_CallContract_0x110101156e8F0743948B2A61aFcf3994A8Fb172e 5  <- stickers 

total                                                       437
total tx detection:                                         66
```

**Pull request merged**